README.rst
sprockets.logging
=================
Making logs nicer since 2015!

|Version| |Downloads| |Travis| |CodeCov| |ReadTheDocs|

Installation
------------
``sprockets.logging`` is available on the
`Python Package Index <https://pypi.python.org/pypi/sprockets.logging>`_
and can be installed via ``pip`` or ``easy_install``:

.. code-block:: bash

   pip install sprockets.logging

Documentation
-------------
https://sprocketslogging.readthedocs.org

Requirements
------------
-  No external requirements

Example
-------
This examples demonstrates the most basic usage of ``sprockets.logging``

.. code-block:: python

   import logging
   import sys

   import sprockets.logging


   formatter = logging.Formatter('%(levelname)s %(message)s {%(context)s}')
   handler = logging.StreamHandler(sys.stdout)
   handler.setFormatter(formatter)
   handler.addFilter(sprockets.logging.ContextFilter(properties=['context']))
   logging.Logger.root.addHandler(handler)
   logging.Logger.root.setLevel(logging.DEBUG)

   # Outputs: INFO Hi there {None}
   logging.info('Hi there')

   # Outputs: INFO No KeyError {bah}
   logging.info('No KeyError', extra={'context': 'bah'})

   # Outputs: INFO Now with context! {foo}
   adapted = logging.LoggerAdapter(logging.Logger.root, extra={'context': 'foo'})
   adapted.info('Now with context!')

Source
------
``sprockets.logging`` source is available on Github at `https://github.com/sprockets/sprockets.logging <https://github.com/sprockets/sprockets.logging>`_

License
-------
``sprockets.logging`` is released under the `3-Clause BSD license <https://github.com/sprockets/sprockets.logging/blob/master/LICENSE>`_.


.. |Version| image:: https://badge.fury.io/py/sprockets.logging.svg?
   :target: http://badge.fury.io/py/sprockets.logging

.. |Travis| image:: https://travis-ci.org/sprockets/sprockets.logging.svg?branch=master
   :target: https://travis-ci.org/sprockets/sprockets.logging

.. |CodeCov| image:: http://codecov.io/github/sprockets/sprockets.logging/coverage.svg?branch=master
   :target: https://codecov.io/github/sprockets/sprockets.logging?branch=master

.. |Downloads| image:: https://pypip.in/d/sprockets.logging/badge.svg?
   :target: https://pypi.python.org/pypi/sprockets.logging

.. |ReadTheDocs| image:: https://readthedocs.org/projects/sprocketslogging/badge/
   :target: https://sprocketslogging.readthedocs.org


docs/history.rst
Version History
===============

`Next Release`_
---------------
- Updated ``tornado_log_function`` to work with Tornado 4.3.

`1.3.2`_ Oct  2, 2015
---------------------
- Switch to packaging as a package instead of a py_module.

`1.3.1`_ Sep 14, 2015
---------------------
- Fix query_arguments handling in Python 3

`1.3.0`_ Aug 28, 2015
---------------------
- Add the traceback and environment if set

`1.2.1`_ Jun 24, 2015
---------------------
- Fix a potential ``KeyError`` when a HTTP request object is not present.

`1.2.0`_ Jun 23, 2015
---------------------
 - Monkeypatch logging.currentframe
 - Include a logging message if it's there

`1.1.0`_ Jun 18, 2015
---------------------
 - Added :class:`sprockets.logging.JSONRequestFormatter`
 - Added :func:`sprockets.logging.tornado_log_function`
 - Added convenience constants and methods as a pass through to Python's logging package:

  - :data:`sprockets.logging.DEBUG` to :data:`logging.DEBUG`
  - :data:`sprockets.logging.ERROR` to :data:`logging.ERROR`
  - :data:`sprockets.logging.INFO` to :data:`logging.INFO`
  - :data:`sprockets.logging.WARN` to :data:`logging.WARN`
  - :data:`sprockets.logging.WARNING` to :data:`logging.WARNING`
  - :func:`sprockets.logging.dictConfig` to :func:`logging.config.dictConfig`
  - :func:`sprockets.logging.getLogger` to :func:`logging.getLogger`

`1.0.0`_ Jun 09, 2015
---------------------
 - Added :class:`sprockets.logging.ContextFilter`

.. _Next Release: https://github.com/sprockets/sprockets.logging/compare/1.3.2...master

.. _1.3.2: https://github.com/sprockets/sprockets.logging/compare/1.3.1...1.3.2
.. _1.3.1: https://github.com/sprockets/sprockets.logging/compare/1.3.0...1.3.1
.. _1.3.0: https://github.com/sprockets/sprockets.logging/compare/1.2.1...1.3.0
.. _1.2.1: https://github.com/sprockets/sprockets.logging/compare/1.2.0...1.2.1
.. _1.2.0: https://github.com/sprockets/sprockets.logging/compare/1.1.0...1.2.0
.. _1.1.0: https://github.com/sprockets/sprockets.logging/compare/1.0.0...1.1.0
.. _1.0.0: https://github.com/sprockets/sprockets.logging/compare/0.0.0...1.0.0


tests.py
import json
import logging
import os
import unittest
import uuid

from tornado import web, testing

import sprockets.logging


def setup_module():
    os.environ.setdefault('ENVIRONMENT', 'development')


class SimpleHandler(web.RequestHandler):

    def get(self):
        if self.get_query_argument('runtime_error', default=None):
            raise RuntimeError(self.get_query_argument('runtime_error'))
        if self.get_query_argument('status_code', default=None) is not None:
            self.set_status(int(self.get_query_argument('status_code')))
        else:
            self.set_status(204)


class RecordingHandler(logging.Handler):

    def __init__(self):
        super(RecordingHandler, self).__init__()
        self.emitted = []

    def emit(self, record):
        self.emitted.append((record, self.format(record)))


class TornadoLoggingTestMixin(object):

    def setUp(self):
        super(TornadoLoggingTestMixin, self).setUp()
        self.access_log = logging.getLogger('tornado.access')
        self.app_log = logging.getLogger('tornado.application')
        self.gen_log = logging.getLogger('tornado.general')
        for logger in (self.access_log, self.app_log, self.gen_log):
            logger.disabled = False

        self.recorder = RecordingHandler()
        root_logger = logging.getLogger()
        root_logger.addHandler(self.recorder)

    def tearDown(self):
        super(TornadoLoggingTestMixin, self).tearDown()
        logging.getLogger().removeHandler(self.recorder)


class TornadoLogFunctionTests(TornadoLoggingTestMixin,
                              testing.AsyncHTTPTestCase):

    def get_app(self):
        return web.Application(
            [web.url('/', SimpleHandler)],
            log_function=sprockets.logging.tornado_log_function)

    @property
    def access_record(self):
        for record, _ in self.recorder.emitted:
            if record.name == 'tornado.access':
                return record

    def test_that_redirect_logged_as_info(self):
        self.fetch('?status_code=303')
        self.assertEqual(self.access_record.levelno, logging.INFO)

    def test_that_client_error_logged_as_warning(self):
        self.fetch('?status_code=400')
        self.assertEqual(self.access_record.levelno, logging.WARNING)

    def test_that_exception_is_logged_as_error(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertEqual(self.access_record.levelno, logging.ERROR)

    def test_that_log_includes_correlation_id(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertIn('correlation_id', self.access_record.args)

    def test_that_log_includes_duration(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertIn('duration', self.access_record.args)

    def test_that_log_includes_headers(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertIn('headers', self.access_record.args)

    def test_that_log_includes_method(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertEqual(self.access_record.args['method'], 'GET')

    def test_that_log_includess_path(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertEqual(self.access_record.args['path'], '/')

    def test_that_log_includes_protocol(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertEqual(self.access_record.args['protocol'], 'http')

    def test_that_log_includes_query_arguments(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertEqual(self.access_record.args['query_args'],
                         {'runtime_error': ['something bad happened']})

    def test_that_log_includes_remote_ip(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertIn('remote_ip', self.access_record.args)

    def test_that_log_includes_status_code(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertEqual(self.access_record.args['status_code'], 500)

    def test_that_log_includes_environment(self):
        self.fetch('/?runtime_error=something%20bad%20happened')
        self.assertEqual(self.access_record.args['environment'],
                         os.environ['ENVIRONMENT'])

    def test_that_log_includes_correlation_id_from_header(self):
        cid = str(uuid.uuid4())
        self.fetch('/?runtime_error=something%20bad%20happened',
                   headers={'Correlation-ID': cid})
        self.assertEqual(self.access_record.args['correlation_id'], cid)


class JSONFormatterTests(TornadoLoggingTestMixin, testing.AsyncHTTPTestCase):

    def setUp(self):
        super(JSONFormatterTests, self).setUp()
        self.recorder.setFormatter(sprockets.logging.JSONRequestFormatter())

    def get_app(self):
        return web.Application(
            [web.url('/', SimpleHandler)],
            log_function=sprockets.logging.tornado_log_function)

    def get_log_line(self, log_name):
        for record, line in self.recorder.emitted:
            if record.name == log_name:
                return json.loads(line)

    def test_that_messages_are_json_encoded(self):
        self.fetch('/')
        for record, line in self.recorder.emitted:
            json.loads(line)

    def test_that_exception_has_traceback(self):
        self.fetch('/?runtime_error=foo')
        entry = self.get_log_line('tornado.application')
        self.assertIsNotNone(entry.get('traceback'))
        self.assertNotEqual(entry['traceback'], [])

    def test_that_successes_do_not_have_traceback(self):
        self.fetch('/')
        for record, line in self.recorder.emitted:
            entry = json.loads(line)
            self.assertNotIn('traceback', entry)


class ContextFilterTests(TornadoLoggingTestMixin, unittest.TestCase):

    def setUp(self):
        super(ContextFilterTests, self).setUp()
        self.logger = logging.getLogger('test-logger')
        self.recorder.setFormatter(
            logging.Formatter('%(message)s {CID %(correlation_id)s}'))
        self.recorder.addFilter(sprockets.logging.ContextFilter(
            properties=['correlation_id']))

    def test_that_property_is_set_to_none_by_filter_when_missing(self):
        self.logger.error('error message')
        _, line = self.recorder.emitted[0]
        self.assertEqual(line, 'error message {CID None}')

    def test_that_extras_property_is_used(self):
        self.logger.error('error message',
                          extra={'correlation_id': 'CORRELATION-ID'})
        _, line = self.recorder.emitted[0]
        self.assertEqual(line, 'error message {CID CORRELATION-ID}')

    def test_that_property_from_logging_adapter_works(self):
        cid = uuid.uuid4()
        logger = logging.LoggerAdapter(self.logger, {'correlation_id': cid})
        logger.error('error message')
        _, line = self.recorder.emitted[0]
        self.assertEqual(line, 'error message {CID %s}' % cid)


sprockets/logging.py
"""
Make good log output easier.

- :class:`ContextFilter` adds fixed properties to a log record
- :class:`JSONRequestFormatter` formats log records as JSON output
- :method:`tornado_log_function` is for use as the
    :class`tornado.web.Application.log_function` in conjunction with
    :class:`JSONRequestFormatter` to output log lines as JSON.

"""
from __future__ import absolute_import

from logging import config
import json
import logging
import os
import sys
import traceback

try:
    from tornado import escape, log
except ImportError:  # pragma no cover
    escape = None
    log = None

version_info = (1, 3, 0)
__version__ = '.'.join(str(v) for v in version_info)

# Shortcut methods and constants to avoid needing to import logging directly
dictConfig = config.dictConfig
getLogger = logging.getLogger

DEBUG = logging.DEBUG
INFO = logging.INFO
WARN = logging.WARN
WARNING = logging.WARNING
ERROR = logging.ERROR


class ContextFilter(logging.Filter):
    """
    Ensures that properties exist on a LogRecord.

    :param list|None properties: optional list of properties that
        will be added to LogRecord instances if they are missing

    This filter implementation will ensure that a set of properties
    exists on every log record which means that you can always refer
    to custom properties in a format string.  Without this, referring
    to a property that is not explicitly passed in will result in an
    ugly ``KeyError`` exception.

    """

    def __init__(self, name='', properties=None):
        logging.Filter.__init__(self, name)
        self.properties = list(properties) if properties else []

    def filter(self, record):
        for property_name in self.properties:
            if not hasattr(record, property_name):
                setattr(record, property_name, None)
        return True


class JSONRequestFormatter(logging.Formatter):
    """Instead of spitting out a "human readable" log line, this outputs
    the log data as JSON.

    """

    def extract_exc_record(self, typ, val, tb):
        """Create a JSON representation of the traceback given the records
        exc_info

        :param `Exception` typ: Exception type of the exception being handled
        :param `Exception` instance val: instance of the Exception class
        :param `traceback` tb: traceback object with the call stack

        :rtype: dict

        """
        exc_record = {'type': typ.__name__,
                      'message': str(val),
                      'stack': []}
        for file_name, line_no, func_name, txt in traceback.extract_tb(tb):
            exc_record['stack'].append({'file': file_name,
                                        'line': str(line_no),
                                        'func': func_name,
                                        'text': txt})
        return exc_record

    def format(self, record):
        """Return the log data as JSON

        :param record logging.LogRecord: The record to format
        :rtype: str

        """
        if hasattr(record, 'exc_info'):
            try:
                traceback = self.extract_exc_record(*record.exc_info)
            except:
                traceback = None

        output = {'name': record.name,
                  'module': record.module,
                  'message': record.msg % record.args,
                  'level': logging.getLevelName(record.levelno),
                  'line_number': record.lineno,
                  'process': record.processName,
                  'timestamp': self.formatTime(record),
                  'thread': record.threadName,
                  'file': record.filename,
                  'request': record.args,
                  'traceback': traceback}
        for key, value in list(output.items()):
            if not value:
                del output[key]
        if 'message' in output:
            output.pop('request', None)
        return json.dumps(output)


def tornado_log_function(handler):
    """Assigned when creating a :py:class:`tornado.web.Application` instance
    by passing the method as the ``log_function`` argument:

    .. code:: python

        app = tornado.web.Application([('/', RequestHandler)],
                                      log_function=tornado_log_function)

    :type handler: :py:class:`tornado.web.RequestHandler`

    """
    status_code = handler.get_status()
    if status_code < 400:
        log_method = log.access_log.info
    elif status_code < 500:
        log_method = log.access_log.warning
    else:
        log_method = log.access_log.error
    correlation_id = (getattr(handler, 'correlation_id', None) or
                      handler.request.headers.get('Correlation-ID', None))
    log_method('', {'correlation_id': correlation_id,
                    'duration': 1000.0 * handler.request.request_time(),
                    'headers': handler.request.headers,
                    'method': handler.request.method,
                    'path': handler.request.path,
                    'protocol': handler.request.protocol,
                    'query_args': escape.recursive_unicode(
                        handler.request.query_arguments),
                    'remote_ip': handler.request.remote_ip,
                    'status_code': status_code,
                    'environment': os.environ.get('ENVIRONMENT')})


def currentframe():
    """Return the frame object for the caller's stack frame."""
    try:
        raise Exception
    except:
        traceback = sys.exc_info()[2]
        frame = traceback.tb_frame
        while True:
            if hasattr(frame, 'f_code'):
                filename = frame.f_code.co_filename
                if filename.endswith('logging.py') or \
                        filename.endswith('logging/__init__.py'):
                    frame = frame.f_back
                    continue
                return frame
        return traceback.tb_frame.f_back

# Monkey-patch currentframe
logging.currentframe = currentframe


examples/tornado-app.py
import logging.config
import signal
import uuid

from tornado import ioloop, web
import sprockets.logging


LOG_CONFIG = {
   'version': 1,
   'handlers': {
      'console': {
         'class': 'logging.StreamHandler',
         'stream': 'ext://sys.stdout',
         'formatter': 'simple',
         'filters': ['context']
      }
   },
   'formatters': {
      'simple': {
         '()': sprockets.logging.JSONRequestFormatter
      }
   },
   'filters': {
      'context': {
         '()': 'sprockets.logging.ContextFilter',
         'properties': ['context']
      }
   },
   'loggers': {
      'tornado': {
         'level': 'DEBUG'
      }
   },
   'root': {
      'handlers': ['console'],
      'level': 'DEBUG'
   },
   'incremental': False
}


class RequestHandler(web.RequestHandler):

   def __init__(self, *args, **kwargs):
      self.parent_log = kwargs.pop('parent_log')
      super(RequestHandler, self).__init__(*args, **kwargs)

   def prepare(self):
      uniq_id = self.request.headers.get('X-UniqID', uuid.uuid4().hex)
      self.logger = logging.LoggerAdapter(
         self.parent_log.getChild('RequestHandler'),
         extra={'context': uniq_id})

   def get(self, object_id):
      self.logger.debug('fetchin %s', object_id)
      self.set_status(200)
      return self.finish()

def sig_handler(signo, frame):
   logging.info('caught signal %d, stopping IO loop', signo)
   iol = ioloop.IOLoop.instance()
   iol.add_callback_from_signal(iol.stop)

if __name__ == '__main__':
   logging.config.dictConfig(LOG_CONFIG)
   logger = logging.getLogger('app')
   app = web.Application([
      web.url('/(?P<object_id>\w+)', RequestHandler,
              kwargs={'parent_log': logger}),
   ], log_function=sprockets.logging.tornado_log_function)
   app.listen(8000)
   signal.signal(signal.SIGINT, sig_handler)
   signal.signal(signal.SIGTERM, sig_handler)
   ioloop.IOLoop.instance().start()
   logger.info('IO loop stopped, exiting')


sprockets/logging/__init__.py
"""
Make good log output easier.

- :class:`ContextFilter` adds fixed properties to a log record
- :class:`JSONRequestFormatter` formats log records as JSON output
- :method:`tornado_log_function` is for use as the
    :class`tornado.web.Application.log_function` in conjunction with
    :class:`JSONRequestFormatter` to output log lines as JSON.

"""
from __future__ import absolute_import

from logging import config
import json
import logging
import os
import sys
import traceback

try:
    from tornado import escape, log
except ImportError:  # pragma no cover
    escape = None
    log = None

version_info = (1, 3, 2)
__version__ = '.'.join(str(v) for v in version_info)

# Shortcut methods and constants to avoid needing to import logging directly
dictConfig = config.dictConfig
getLogger = logging.getLogger

DEBUG = logging.DEBUG
INFO = logging.INFO
WARN = logging.WARN
WARNING = logging.WARNING
ERROR = logging.ERROR


class ContextFilter(logging.Filter):
    """
    Ensures that properties exist on a LogRecord.

    :param list|None properties: optional list of properties that
        will be added to LogRecord instances if they are missing

    This filter implementation will ensure that a set of properties
    exists on every log record which means that you can always refer
    to custom properties in a format string.  Without this, referring
    to a property that is not explicitly passed in will result in an
    ugly ``KeyError`` exception.

    """

    def __init__(self, name='', properties=None):
        logging.Filter.__init__(self, name)
        self.properties = list(properties) if properties else []

    def filter(self, record):
        for property_name in self.properties:
            if not hasattr(record, property_name):
                setattr(record, property_name, None)
        return True


class JSONRequestFormatter(logging.Formatter):
    """Instead of spitting out a "human readable" log line, this outputs
    the log data as JSON.

    """

    def extract_exc_record(self, typ, val, tb):
        """Create a JSON representation of the traceback given the records
        exc_info

        :param `Exception` typ: Exception type of the exception being handled
        :param `Exception` instance val: instance of the Exception class
        :param `traceback` tb: traceback object with the call stack

        :rtype: dict

        """
        exc_record = {'type': typ.__name__,
                      'message': str(val),
                      'stack': []}
        for file_name, line_no, func_name, txt in traceback.extract_tb(tb):
            exc_record['stack'].append({'file': file_name,
                                        'line': str(line_no),
                                        'func': func_name,
                                        'text': txt})
        return exc_record

    def format(self, record):
        """Return the log data as JSON

        :param record logging.LogRecord: The record to format
        :rtype: str

        """
        if hasattr(record, 'exc_info'):
            try:
                traceback = self.extract_exc_record(*record.exc_info)
            except:
                traceback = None

        output = {'name': record.name,
                  'module': record.module,
                  'message': record.msg % record.args,
                  'level': logging.getLevelName(record.levelno),
                  'line_number': record.lineno,
                  'process': record.processName,
                  'timestamp': self.formatTime(record),
                  'thread': record.threadName,
                  'file': record.filename,
                  'request': record.args,
                  'traceback': traceback}
        for key, value in list(output.items()):
            if not value:
                del output[key]
        if 'message' in output:
            output.pop('request', None)
        return json.dumps(output)


def tornado_log_function(handler):
    """Assigned when creating a :py:class:`tornado.web.Application` instance
    by passing the method as the ``log_function`` argument:

    .. code:: python

        app = tornado.web.Application([('/', RequestHandler)],
                                      log_function=tornado_log_function)

    :type handler: :py:class:`tornado.web.RequestHandler`

    """
    status_code = handler.get_status()
    if status_code < 400:
        log_method = log.access_log.info
    elif status_code < 500:
        log_method = log.access_log.warning
    else:
        log_method = log.access_log.error
    correlation_id = (getattr(handler, 'correlation_id', None) or
                      handler.request.headers.get('Correlation-ID', None))
    log_method('', {'correlation_id': correlation_id,
                    'duration': 1000.0 * handler.request.request_time(),
                    'headers': dict(handler.request.headers),
                    'method': handler.request.method,
                    'path': handler.request.path,
                    'protocol': handler.request.protocol,
                    'query_args': escape.recursive_unicode(
                        handler.request.query_arguments),
                    'remote_ip': handler.request.remote_ip,
                    'status_code': status_code,
                    'environment': os.environ.get('ENVIRONMENT')})


def currentframe():
    """Return the frame object for the caller's stack frame."""
    try:
        raise Exception
    except:
        traceback = sys.exc_info()[2]
        frame = traceback.tb_frame
        while True:
            if hasattr(frame, 'f_code'):
                filename = frame.f_code.co_filename
                if filename.endswith('logging.py') or \
                        filename.endswith('logging/__init__.py'):
                    frame = frame.f_back
                    continue
                return frame
        return traceback.tb_frame.f_back

# Monkey-patch currentframe
logging.currentframe = currentframe
